File: Components/archive_statistic.m
function [output1,output2] = archive_statistic(varargin)
% Collect the average and standard deviation of solution's fitness of each
% iteration.

mode = varargin{end};
switch mode
    case 'execute'
        Solution    = varargin{1};
        CurrArchive = varargin{2};

        Fitness = Solution.fits;
        output1 = [mean(Fitness),std(Fitness)];
        output1 = [CurrArchive;output1];
    
    case 'parameter'
        % no parameter

    case 'behavior'
        output1 = {'';''}; 
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: src/autooptlib/components/archive_statistic.py
"""Track mean and std of fitness."""
from __future__ import annotations

from typing import Any, Sequence

import numpy as np

from ._utils import flex_get


def archive_statistic(*args):
    mode = args[-1]
    if mode == "execute":
        solution = args[0]
        archive = args[1] if len(args) > 1 else []

        fits = flex_get(solution, "fits")
        if callable(fits):
            values = np.asarray(fits()).reshape(-1)
        else:
            values = np.asarray(fits).reshape(-1)
        stat = np.array([[np.mean(values), np.std(values)]])
        if archive is None or len(archive) == 0:
            return stat, None
        archive_arr = np.asarray(archive)
        return np.vstack([archive_arr, stat]), None

    if mode == "parameter":
        return None, None

    if mode == "behavior":
        return ["", ""], None

    raise ValueError(f"Unsupported mode: {mode}")